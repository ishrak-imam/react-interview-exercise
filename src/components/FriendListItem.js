import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import styles from './FriendListItem.css'
import { connect } from 'react-redux'

class FriendListItem extends Component {
  render () {
    const { friendId, friendlist: { friends } } = this.props
    const { name, gender, id, starred } = friends[friendId]

    return (
      <li className={styles.friendListItem}>
        <div className={styles.friendInfos}>
          <div><span>{name}</span></div>
          <div>
            {/* <small>xx friends in common</small> */}
            <i className={classnames('fa', {
              'fa-female': gender === 'F',
              'fa-male': gender === 'M'
            })} />
          </div>
        </div>
        <div className={styles.friendActions}>
          <button className={`btn btn-default ${styles.btnAction}`}
            onClick={() => this.props.starFriend(id)}>
            <i className={classnames('fa', {
              'fa-star': starred,
              'fa-star-o': !starred
            })} />
          </button>
          <button className={`btn btn-default ${styles.btnAction}`}
            onClick={() => this.props.deleteFriend(id)}>
            <i className='fa fa-trash' />
          </button>
        </div>
      </li>
    )
  }
}

FriendListItem.propTypes = {
  friendId: PropTypes.string.isRequired,
  starFriend: PropTypes.func.isRequired
}

function mapStateToProps (state) {
  return state
}

export default connect(mapStateToProps, null)(FriendListItem)
