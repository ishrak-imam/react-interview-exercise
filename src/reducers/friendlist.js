import * as types from '../constants/ActionTypes'

const ITEMS_PER_PAGE = 3

const data = [
  {
    name: 'Theodore Roosevelt',
    starred: true,
    id: 0,
    gender: 'M'
  },
  {
    name: 'Abraham Lincoln',
    starred: false,
    id: 1,
    gender: 'M'
  },
  {
    name: 'George Washington',
    starred: false,
    id: 2,
    gender: 'M'
  }
]

const normalizeList = list => {
  return list.reduce((obj, item, index) => {
    obj.map[item.id] = item
    obj.ids.push(index)
    return obj
  }, { map: {}, ids: [] })
}

const chunkList = (list, size) => {
  let chunkedList = []
  const noOfChunks = Math.ceil(list.length / size)
  for (let i = 0; i < noOfChunks; i++) {
    chunkedList.push(list.slice(i * size, (i + 1) * size))
  }
  return chunkedList
}

const getPages = ids => {
  const chunkedIds = chunkList(ids, ITEMS_PER_PAGE)
  return chunkedIds.reduce((pages, chunk, index) => {
    pages[index + 1] = { ids: chunk }
    return pages
  }, {})
}

const { map, ids } = normalizeList(data)

const initialState = {
  friends: map,
  ids,
  pagination: {
    itemsPerPage: ITEMS_PER_PAGE,
    currentPage: '1',
    pages: getPages(ids)
  }
}

export default function friends (state = initialState, action) {
  switch (action.type) {
    case types.ADD_FRIEND:
      let ids = state.ids
      let friends = { ...state.friends, [ids.length]: { ...action.friend, id: ids.length } }
      ids.push(ids.length)

      return {
        ...state,
        friends,
        ids,
        pagination: {
          ...state.pagination,
          pages: getPages(ids)
        }
      }

    case types.DELETE_FRIEND:
      ids = state.ids.filter(id => id !== action.id)
      friends = Object.assign({}, state.friends)
      delete friends[action.id]

      // let { [action.id]: deleted, ...afterDelete } = friends

      return {
        ...state,
        friends,
        ids,
        pagination: {
          ...state.pagination,
          pages: getPages(ids)
        }
      }

    case types.STAR_FRIEND:
      friends = state.friends
      let friend = friends[action.id]
      return {
        ...state,
        friends: {
          ...friends,
          [action.id]: {
            ...friend,
            starred: true
          }
        }
      }

    case types.GO_TO_PAGE:
      return {
        ...state,
        pagination: {
          ...state.pagination,
          currentPage: action.page
        }
      }

    default:
      return state
  }
}
