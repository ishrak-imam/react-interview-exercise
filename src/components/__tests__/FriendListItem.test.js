import React from 'react'
import { create } from 'react-test-renderer'
import { FriendListItem } from '../FriendListItem'

import { initialState } from '../../reducers/friendlist'

test('FriendListItem component', () => {
  const component = create(
    <FriendListItem
      friendId='1'
      deleteFriend={() => {}}
      starFriend={() => {}}
      friendlist={initialState}
    />
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
