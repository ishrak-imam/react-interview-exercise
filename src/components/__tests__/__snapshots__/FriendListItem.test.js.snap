exports[`test FriendListItem component 1`] = `
<li
  className={undefined}>
  <div
    className={undefined}>
    <div>
      <span>
        Abraham Lincoln
      </span>
    </div>
    <div>
      <i
        className="fa undefined fa-male" />
    </div>
  </div>
  <div
    className={undefined}>
    <button
      className="btn btn-default undefined"
      onClick={[Function]}>
      <i
        className="fa fa-star-o" />
    </button>
    <button
      className="btn btn-default undefined"
      onClick={[Function]}>
      <i
        className="fa fa-trash" />
    </button>
  </div>
</li>
`;
